frame/pallet-account/src/lib.rs

```diff
@@ -0,0 +1,203 @@
+    use frame_support::{
+        dispatch::{Dispatchable, GetDispatchInfo},
+        traits::IsSubType,
+    };
+    use frame_system::pallet_prelude::*;
+    use sp_std::prelude::*;
+
+    /// The current storage version.                                                                                      
+    const STORAGE_VERSION: StorageVersion = StorageVersion::new(1);
+
+    #[pallet::pallet]
+    #[pallet::generate_store(pub(super) trait Store)]
+    #[pallet::storage_version(STORAGE_VERSION)]
+    pub struct Pallet<T>(PhantomData<T>);
+
+    #[pallet::config]
+    pub trait Config: frame_system::Config {
+        /// Custom origin type.
+        type CustomOrigin: Parameter + TryInto<Self::AccountId> + MaxEncodedLen;
+        /// Parameter that defin different origin options and how to create it.
+        type CustomOriginKind: Parameter + OriginDeriving<Self::AccountId, Self::CustomOrigin>;
+        /// The runtime origin type.
+        type RuntimeOrigin: From<Self::CustomOrigin>
+            + From<frame_system::RawOrigin<Self::AccountId>>;
+        /// The overarching call type.
+        type RuntimeCall: Parameter
+            + Dispatchable<RuntimeOrigin = <Self as Config>::RuntimeOrigin>
+            + GetDispatchInfo
+            + From<frame_system::Call<Self>>
+            + IsSubType<Call<Self>>
+            + IsType<<Self as frame_system::Config>::RuntimeCall>;
+        /// General event type.
+        type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+        /// Weight information for extrinsics in this pallet.
+        type WeightInfo: WeightInfo;
+    }
+
+    #[pallet::error]
+    pub enum Error<T> {
+        /// Origin with given index not registered.
+        UnregisteredOrigin,
+    }
+
+    #[pallet::event]
+    #[pallet::generate_deposit(pub(crate) fn deposit_event)]
+    pub enum Event<T: Config> {
+        NewOrigin {
+            account: T::AccountId,
+            origin: T::CustomOrigin,
+        },
+        ProxyCall {
+            origin: T::CustomOrigin,
+            result: DispatchResult,
+        },
+    }
+
+    #[pallet::origin]
+    pub type Origin<T> = <T as Config>::CustomOrigin;
+
+    /// Account origins
+    #[pallet::storage]
+    pub type AccountOrigin<T: Config> =
+        StorageDoubleMap<_, Blake2_128Concat, T::AccountId, Twox64Concat, u32, T::CustomOrigin>;
+
+    /// Account last origin index
+    #[pallet::storage]
+    pub type AccountLastOrigin<T: Config> =
+        StorageMap<_, Twox64Concat, T::AccountId, u32, ValueQuery>;
+
+    #[pallet::call]
+    impl<T: Config> Pallet<T> {
+        /// Derive new origin for account.
+        ///
+        /// The dispatch origin for this call must be _Signed_.
+        #[pallet::weight(
+            T::WeightInfo::new_origin()
+                .saturating_add(T::DbWeight::get().reads_writes(1, 2))
+        )]
+        #[pallet::call_index(0)]
+        pub fn new_origin(
```

Comment:
As the origin derivation algorithm is deterministic, could we skip the origin registering? The derived origin could be derived just in time in a `proxy_call`, like:

```rust
Account::proxy_call(alice, derive_kind, index, call);
```

Reply:
I believe derivation algorithm is just a sample, it not restricted to be deterministic. Account registration helps to prevent spamming and account address manipulation.

Reply:
@akru what is the usecase behind allowing single account to have multiple derived address of the same type?

The way I see it, it just adds unnecessary complexity.
If we only allow one account type to be derived from the origin, then it's much simpler.
If an account wants to control multiple different derived accounts, it can use proxy to achieve that.

Reply:
Shaun also commented above([here](https://github.com/AstarNetwork/astar-frame/pull/130#discussion_r1197707974)) about what happens when account is _killed_. Without any locked or reserved funds, these storage items will remain in the DB even without an owner.